File: csharp/perfect-numbers/PerfectNumbers.cs
using System.Linq;

public class PerfectNumbers
{
    public static NumberType Classify(int number)
    {
        var sum = 
            Enumerable
                .Range(1, number / 2)
                .Where(x => number % x == 0)
                .Sum();

        return DetermineNumberType(number, sum);
    }

    private static NumberType DetermineNumberType(int number, int sum)
    {
        if (sum == number)
        {
            return NumberType.Perfect;
        }
        return sum > number ? NumberType.Abundant : NumberType.Deficient;
    }
}

File: swift/perfect-numbers/PerfectNumbers.swift
enum Classification {
    case abundant
    case deficient
    case perfect
}

class NumberClassifier {
    let number : Int

    init(number number: Int) {
        self.number = number
    }

    var classification : Classification {
        let sum = (1..<number).filter({ number % $0 == 0}).reduce(0, +)

        switch (sum) {
            case _ where sum == number:
                return .perfect
            case _ where sum > number:
                return .abundant
            default:
                return .deficient
        }
    }
}

